File: Design Patterns/3.Builder/builder.py
from enum import Enum
from abc import ABC, abstractmethod

class Starter(Enum):
    SALAD = 1
    SOUP = 2
    BRUSCHETTA = 3
    VEGGIE_STICKS = 4
    CHICKEN_WINGS = 5

class Main(Enum):
    GRILLED_CHICKEN = 1
    PASTA = 2
    VEGGIE_STIR_FRY = 3
    FISH = 4
    PIZZA = 5

class Dessert(Enum):
    FRUIT_SALAD = 1
    ICE_CREAM = 2
    CHOCOLATE_CAKE = 3
    VEGAN_PUDDING = 4
    CHEESECAKE = 5

class Drink(Enum):
    WATER = 1
    VEGAN_SHAKE = 2
    SODA = 3
    FRUIT_JUICE = 4

class Meal:
    def __init__(self):
        self._starter = None
        self._main = None
        self._dessert = None
        self._drink = None

    @property
    def starter(self):
        return self._starter

    @property
    def main(self):
        return self._main

    @property
    def dessert(self):
        return self._dessert

    @property
    def drink(self):
        return self._drink

    @starter.setter
    def starter(self, starter):
        self._starter = starter

    @main.setter
    def main(self, main):
        self._main = main

    @dessert.setter
    def dessert(self, dessert):
        self._dessert = dessert

    @drink.setter
    def drink(self, drink):
        self._drink = drink

class Builder(ABC):
    @abstractmethod
    def add_starter(self):
        pass

    @abstractmethod
    def add_main_course(self):
        pass

    @abstractmethod
    def add_dessert(self):
        pass

    @abstractmethod
    def add_drink(self):
        pass

class VeganMealBuilder(Builder):
    def __init__(self):
        self.meal = Meal()

    def add_starter(self):
        self.meal.starter = Starter.SALAD

    def add_main_course(self):
        self.meal.main = Main.VEGGIE_STIR_FRY

    def add_dessert(self):
        self.meal.dessert = Dessert.VEGAN_PUDDING

    def add_drink(self):
        self.meal.drink = Drink.VEGAN_SHAKE

    def build(self):
        return self.meal

class HealthyMealBuilder(Builder):
    def __init__(self):
        self.meal = Meal()

    def add_starter(self):
        self.meal.starter = Starter.SALAD

    def add_main_course(self):
        self.meal.main = Main.GRILLED_CHICKEN

    def add_dessert(self):
        self.meal.dessert = Dessert.FRUIT_SALAD

    def add_drink(self):
        self.meal.drink = Drink.WATER

    def build(self):
        return self.meal

class Director:
    def construct_vegan_meal(self, builder):
        builder.add_starter()
        builder.add_main_course()
        builder.add_dessert()
        builder.add_drink()

    def construct_healthy_meal(self, builder):
        builder.add_starter()
        builder.add_main_course()
        builder.add_dessert()
        builder.add_drink()

def testBuilder():
    director = Director()
    vegan_builder = VeganMealBuilder()
    director.construct_vegan_meal(vegan_builder)

    vegan_meal = vegan_builder.build()
    print("Vegan Meal constructed: ")
    print(f"Starter: {vegan_meal.starter.name}")
    print(f"Main: {vegan_meal.main.name}")
    print(f"Dessert: {vegan_meal.dessert.name}")
    print(f"Drink: {vegan_meal.drink.name}")

    healthy_builder = HealthyMealBuilder()
    director.construct_healthy_meal(healthy_builder)
    healthy_meal = healthy_builder.build()
    print("Healthy Meal constructed: ")
    print(f"Starter: {healthy_meal.starter.name}")
    print(f"Main: {healthy_meal.main.name}")
    print(f"Dessert: {healthy_meal.dessert.name}")
    print(f"Drink: {healthy_meal.drink.name}")

if __name__ == "__main__":
    testBuilder()

File: Design Patterns/3.Builder/builder.ts
enum Starter {
  SALAD = 'SALAD',
  SOUP = 'SOUP',
  BRUSCHETTA = 'BRUSCHETTA',
  VEGGIE_STICKS = 'VEGGIE_STICKS',
  CHICKEN_WINGS = 'CHICKEN_WINGS',
}

enum Main {
  GRILLED_CHICKEN = 'GRILLED_CHICKEN',
  PASTA = 'PASTA',
  VEGGIE_STIR_FRY = 'VEGGIE_STIR_FRY',
  FISH = 'FISH',
  PIZZA = 'PIZZA',
}

enum Dessert {
  FRUIT_SALAD = 'FRUIT_SALAD',
  ICE_CREAM = 'ICE_CREAM',
  CHOCOLATE_CAKE = 'CHOCOLATE_CAKE',
  VEGAN_PUDDING = 'VEGAN_PUDDING',
  CHEESECAKE = 'CHEESECAKE',
}

enum Drink {
  WATER = 'WATER',
  VEGAN_SHAKE = 'VEGAN_SHAKE',
  SODA = 'SODA',
  FRUIT_JUICE = 'FRUIT_JUICE',
}

class Meal {
  private _starter: Starter | null = null;
  private _main: Main | null = null;
  private _dessert: Dessert | null = null;
  private _drink: Drink | null = null;

  get starter(): Starter | null {
    return this._starter;
  }
  get main(): Main | null {
    return this._main;
  }
  get dessert(): Dessert | null {
    return this._dessert;
  }
  get drink(): Drink | null {
    return this._drink;
  }

  set starter(value: Starter | null) {
    this._starter = value;
  }
  set main(value: Main | null) {
    this._main = value;
  }
  set dessert(value: Dessert | null) {
    this._dessert = value;
  }
  set drink(value: Drink | null) {
    this._drink = value;
  }
}

interface Builder {
  addStarter(): this;
  addMainCourse(): this;
  addDessert(): this;
  addDrink(): this;
  build(): Meal;
}

class VeganMealBuilder implements Builder {
  private meal: Meal = new Meal();

  addStarter(): this {
    this.meal.starter = Starter.SALAD;
    return this;
  }

  addMainCourse(): this {
    this.meal.main = Main.VEGGIE_STIR_FRY;
    return this;
  }

  addDessert(): this {
    this.meal.dessert = Dessert.VEGAN_PUDDING;
    return this;
  }

  addDrink(): this {
    this.meal.drink = Drink.VEGAN_SHAKE;
    return this;
  }

  build(): Meal {
    if (!this.meal.starter || !this.meal.main) {
      throw new Error('Meal is incomplete');
    }
    return this.meal;
  }
}

class HealthyMealBuilder implements Builder {
  private meal: Meal = new Meal();

  addStarter(): this {
    this.meal.starter = Starter.SALAD;
    return this;
  }

  addMainCourse(): this {
    this.meal.main = Main.GRILLED_CHICKEN;
    return this;
  }

  addDessert(): this {
    this.meal.dessert = Dessert.FRUIT_SALAD;
    return this;
  }

  addDrink(): this {
    this.meal.drink = Drink.WATER;
    return this;
  }

  build(): Meal {
    return this.meal;
  }
}

class Director {
  constructVeganMeal(builder: Builder): Meal {
    return builder.addStarter().addMainCourse().addDessert().addDrink().build();
  }

  constructHealthyMeal(builder: Builder): Meal {
    return builder.addStarter().addMainCourse().addDessert().addDrink().build();
  }

  constructCustomMeal(
    builder: Builder,
    options: {
      withStarter?: boolean;
      withDessert?: boolean;
      withDrink?: boolean;
    }
  ): Meal {
    let meal = builder.addMainCourse();

    if (options.withStarter) meal = meal.addStarter();
    if (options.withDessert) meal = meal.addDessert();
    if (options.withDrink) meal = meal.addDrink();

    return meal.build();
  }
}

function testBuilder(): void {
  const director = new Director();

  const veganMeal = director.constructVeganMeal(new VeganMealBuilder());
  const healthyMeal = director.constructHealthyMeal(new HealthyMealBuilder());

  const customMeal = director.constructCustomMeal(new VeganMealBuilder(), {
    withStarter: true,
    withDrink: true,
  });

  console.log('Meals constructed:', { veganMeal, healthyMeal, customMeal });
}

testBuilder();
